pkg/controller/daemon_controller.go

```diff
@@ -0,0 +1,432 @@
+	// more pods until all the effects (expectations) of a daemon controller's create/delete have been observed.
+	dm.podStore.Store, dm.podController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return dm.kubeClient.Pods(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return dm.kubeClient.Pods(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), rv)
+			},
+		},
+		&api.Pod{},
+		FullDaemonPodResyncPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    dm.addPod,
+			UpdateFunc: dm.updatePod, // TODO: add update function, and call it
+			DeleteFunc: dm.deletePod,
+		},
+	)
+	// Watch for new nodes or updates to nodes - daemons are launched on new nodes, and possibly when labels on nodes change,
+	dm.nodeStore.Store, dm.nodeController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return dm.kubeClient.Nodes().List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return dm.kubeClient.Nodes().Watch(labels.Everything(), fields.Everything(), rv)
+			},
+		},
+		&api.Node{},
+		FullNodeResyncPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    dm.addNode,
+			UpdateFunc: dm.updateNode,
+			DeleteFunc: func(node interface{}) {},
+		},
+	)
+	dm.syncHandler = dm.syncDaemonController
+	return dm
+}
+
+func (dm *DaemonManager) Run(workers int, stopCh <-chan struct{}) {
+	go dm.dcController.Run(stopCh)
+	go dm.podController.Run(stopCh)
+	go dm.nodeController.Run(stopCh)
+	for i := 0; i < workers; i++ {
+		go util.Until(dm.worker, time.Second, stopCh)
+	}
+	<-stopCh
+	glog.Infof("Shutting down Daemon Controller Manager")
+	dm.queue.ShutDown()
+}
+
+func (dm *DaemonManager) worker() {
+	for {
+		func() {
+			key, quit := dm.queue.Get()
+			if quit {
+				return
+			}
+			defer dm.queue.Done(key)
+			err := dm.syncHandler(key.(string))
+			if err != nil {
+				glog.Errorf("Error syncing daemon controller: %v", err)
+			}
+		}()
+	}
+}
+
+func (dm *DaemonManager) enqueueAllControllers() {
+	daemonControllers, err := dm.dcStore.List()
+	if err != nil {
+		glog.Errorf("Error enqueueing daemon controllers: %v", err)
+		return
+	}
+	for i := range daemonControllers {
+		dm.enqueueController(&daemonControllers[i])
+	}
+}
+
+func (dm *DaemonManager) enqueueController(obj interface{}) {
+	key, err := controllerKeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+	dm.queue.Add(key)
+}
+
+func (dm *DaemonManager) getPodDaemonController(pod *api.Pod) *api.DaemonController {
+	controllers, err := dm.dcStore.GetPodDaemonController(pod)
+	if err != nil {
+		glog.V(4).Infof("No controllers found for pod %v, daemon manager will avoid syncing", pod.Name)
```

Comment:
Generally I'd prefer that you return an error instead of logging, and not use a return value of nil to indicate that an error happened.